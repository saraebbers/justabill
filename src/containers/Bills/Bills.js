import React from 'react';

const Bills = () => {
  const congress = 115;

  return(
    <div>
      <h3>
        `Bills signed into law during the ${congress} congress.`
      </h3>
    </div>

  )
}

export default Bills;