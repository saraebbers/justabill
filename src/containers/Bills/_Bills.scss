.cardContainer {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(275px, 1fr));
  grid-column-gap: 10px;
  justify-items: center;
  justify-content: space-around;
  margin: 10px;
  padding: 10px;
}
.bills-pg {
    // height: 100vh;
  background: url(./assets/congressFloor.jpg)
    top center;
  background-size: cover;
}

.bill-msg {
  border: solid #718F94 5px;
  background-color: #ADA380;
  margin-top: 20px;
  padding: 10px;
  width: 90%;
  color: #151E3F;

  h3 {
    text-align: center;
    font-size: 1.5rem;
    padding: 10px;
  }
}

.bill-container {
  justify-content: start;
  display: grid;
  grid-template-columns: 1fr;
  justify-items: center;
  margin: 0;
}

.no-Bills {
  display: grid;
  background-color: #DBCFB0;
  margin: 25px 5px;
  padding: 5px;
  border: solid #718F94 4px;
  color: #420C14;
  height: 150px;
  align-items: start;
  justify-items: center;

  h4 {
    margin-left: 10px;
    text-align: center;
    font-size: 1rem;
    color: #151E3F;
    padding: 5px;
  }

  p {
    margin-left: 10px;
    font-size: .75rem;
  }
}

button {
    justify-self: end;
    padding: 10px;
    background-color: #420C14;
    border: solid #718F94 2px;
    color: #DBCFB0;
    border-radius: 20px;
    outline: none;
    margin: 10px;

    &.active {
      background-color: #718F94;
      border: solid #420C14 2px;
      outline: none;
    }

    &:hover {
      background-color: #718F94;
      border: solid #420C14 2px;
      outline: none;
    }

  }