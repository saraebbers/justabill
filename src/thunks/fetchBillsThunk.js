import { isLoading, hasErrored, getBills } from '../actions/index';
import { apiKey } from '../utils/apiKey'

export const fetchBillsThunk = (url) => {
  return async (dispatch) => {
    try {
      console.log('pre')
      dispatch(isLoading(true))
      console.log('post')
      const response = await fetch(url, {
        method: 'GET',
        headers: {
          'X-API-Key': `${apiKey}`
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      // dispatch(isLoading(false))
      const billsData = await response.json()
      return billsData.results
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}
