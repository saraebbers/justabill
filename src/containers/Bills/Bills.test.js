import React from 'react';
import { shallow } from 'enzyme';
import { mapStateToProps, mapDispatchToProps, Bills } from './Bills';
import { fetchBillsThunk } from '../../thunks/fetchBillsThunk';
jest.mock('../../thunks/fetchBillsThunk')


describe('Bills', () => {
  let mockBillArray;
  let mockErrorMessage;
  let mockFavorites;
  let mockFetchBillsThunk;
  let mockIsLoading;
  let mockLocalState

  beforeEach(() => {
    mockBillArray=[{congress: 115, bills: [{id: 'hr123', name:'sara', somethingelse: 'xxx'}, {id: 'hr234', name:'David', somethingelse: 'xxx'}]}]
    mockFavorites=['hr-45123', 'hr-7653']
    mockErrorMessage = ''
    mockFetchBillsThunk= jest.fn()
    mockIsLoading=false;
    mockLocalState = {congress: 115}
  })

  describe('bills container', () => {

    it('should match the snapshot if is Loading is true', () => {
      mockIsLoading = true
      let state = mockLocalState 
      const wrapper = shallow(<Bills isLoading={mockIsLoading} billArray={mockBillArray} errorMessage={mockErrorMessage} favorites={mockFavorites} fetchBillsThunk={mockFetchBillsThunk} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('should match the snapshot if is Loading is false', () => {
      const wrapper = shallow(<Bills isLoading={mockIsLoading} billArray={mockBillArray} errorMessage={mockErrorMessage} favorites={mockFavorites} fetchBillsThunk={mockFetchBillsThunk} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('should fetch bills with the current value of state when the component mounts', async () => {
       const wrapper = shallow(<Bills isLoading={mockIsLoading} billArray={mockBillArray} errorMessage={mockErrorMessage} favorites={mockFavorites} fetchBillsThunk={mockFetchBillsThunk} />)
       await wrapper.instance().componentDidMount()
       expect(mockFetchBillsThunk).toHaveBeenCalled()
    })

    it('should have the correct default local state', () => {
      const wrapper = shallow(<Bills isLoading={mockIsLoading} billArray={mockBillArray} errorMessage={mockErrorMessage} favorites={mockFavorites} fetchBillsThunk={mockFetchBillsThunk} />)
      const expected = mockLocalState
      expect(wrapper.state()).toEqual(expected)
    })
  })

  describe('mapStateToProps', () => {
    it('should return a props object with the correct information from state parsed', () => {
      const mockState = {
        bills: mockBillArray,
        isLoading: mockIsLoading,
        errorMessage: 'this is an error message',
        favorites: mockFavorites,
        somethingElse: 'something else'
      }
      const expected = {
        billArray: mockBillArray,
        isLoading: mockIsLoading,
        errorMessage: 'this is an error message',
        favorites: mockFavorites
      }
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('calls dispatch with fetchBillsThunk when fetchBillsThunk is called', () => {
      const mockDispatch = jest.fn()      
      const url = 'https://something/something'
      fetchBillsThunk.mockImplementation(() => mockBillArray)
      const actionToDispatch = fetchBillsThunk(url)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.fetchBillsThunk(url)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })

})
