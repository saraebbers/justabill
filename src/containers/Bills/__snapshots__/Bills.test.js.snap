// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bills bills comtainer should match the snapshot if is Loading is false 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Bills
    billArray={
      Array [
        Object {
          "bills": Array [
            Object {
              "id": "hr123",
              "name": "sara",
              "somethingelse": "xxx",
            },
            Object {
              "id": "hr234",
              "name": "David",
              "somethingelse": "xxx",
            },
          ],
          "congress": 115,
        },
      ]
    }
    errorMessage=""
    favorites={
      Array [
        "hr-45123",
        "hr-7653",
      ]
    }
    fetchBillsThunk={
      [MockFunction] {
        "calls": Array [
          Array [
            "https://api.propublica.org/congress/v1/115/both/bills/enacted.json",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    isLoading={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <section
        className="bill-container"
      >
        <div
          className="bill-msg"
        >
          <h3>
            Below are the bills that became law during the 115th Congress.
          </h3>
        </div>
        <div
          className="cardContainer"
        >
          <withRouter(Connect(Card))
            id="hr123"
            name="sara"
            somethingelse="xxx"
          />
          <withRouter(Connect(Card))
            id="hr234"
            name="David"
            somethingelse="xxx"
          />
        </div>
      </section>,
      "className": "bills-pg",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="bill-msg"
          >
            <h3>
              Below are the bills that became law during the 115th Congress.
            </h3>
          </div>,
          <div
            className="cardContainer"
          >
            <withRouter(Connect(Card))
              id="hr123"
              name="sara"
              somethingelse="xxx"
            />
            <withRouter(Connect(Card))
              id="hr234"
              name="David"
              somethingelse="xxx"
            />
          </div>,
        ],
        "className": "bill-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h3>
              Below are the bills that became law during the 115th Congress.
            </h3>,
            "className": "bill-msg",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Below are the bills that became law during the 115th Congress.",
            },
            "ref": null,
            "rendered": "Below are the bills that became law during the 115th Congress.",
            "type": "h3",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <withRouter(Connect(Card))
                id="hr123"
                name="sara"
                somethingelse="xxx"
              />,
              <withRouter(Connect(Card))
                id="hr234"
                name="David"
                somethingelse="xxx"
              />,
            ],
            "className": "cardContainer",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "hr123",
              "nodeType": "function",
              "props": Object {
                "id": "hr123",
                "name": "sara",
                "somethingelse": "xxx",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "hr234",
              "nodeType": "function",
              "props": Object {
                "id": "hr234",
                "name": "David",
                "somethingelse": "xxx",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "section",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <section
          className="bill-container"
        >
          <div
            className="bill-msg"
          >
            <h3>
              Below are the bills that became law during the 115th Congress.
            </h3>
          </div>
          <div
            className="cardContainer"
          >
            <withRouter(Connect(Card))
              id="hr123"
              name="sara"
              somethingelse="xxx"
            />
            <withRouter(Connect(Card))
              id="hr234"
              name="David"
              somethingelse="xxx"
            />
          </div>
        </section>,
        "className": "bills-pg",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="bill-msg"
            >
              <h3>
                Below are the bills that became law during the 115th Congress.
              </h3>
            </div>,
            <div
              className="cardContainer"
            >
              <withRouter(Connect(Card))
                id="hr123"
                name="sara"
                somethingelse="xxx"
              />
              <withRouter(Connect(Card))
                id="hr234"
                name="David"
                somethingelse="xxx"
              />
            </div>,
          ],
          "className": "bill-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h3>
                Below are the bills that became law during the 115th Congress.
              </h3>,
              "className": "bill-msg",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Below are the bills that became law during the 115th Congress.",
              },
              "ref": null,
              "rendered": "Below are the bills that became law during the 115th Congress.",
              "type": "h3",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <withRouter(Connect(Card))
                  id="hr123"
                  name="sara"
                  somethingelse="xxx"
                />,
                <withRouter(Connect(Card))
                  id="hr234"
                  name="David"
                  somethingelse="xxx"
                />,
              ],
              "className": "cardContainer",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "hr123",
                "nodeType": "function",
                "props": Object {
                  "id": "hr123",
                  "name": "sara",
                  "somethingelse": "xxx",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "hr234",
                "nodeType": "function",
                "props": Object {
                  "id": "hr234",
                  "name": "David",
                  "somethingelse": "xxx",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "section",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Bills bills comtainer should match the snapshot if is Loading is true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Bills
    billArray={
      Array [
        Object {
          "bills": Array [
            Object {
              "id": "hr123",
              "name": "sara",
              "somethingelse": "xxx",
            },
            Object {
              "id": "hr234",
              "name": "David",
              "somethingelse": "xxx",
            },
          ],
          "congress": 115,
        },
      ]
    }
    errorMessage=""
    favorites={
      Array [
        "hr-45123",
        "hr-7653",
      ]
    }
    fetchBillsThunk={
      [MockFunction] {
        "calls": Array [
          Array [
            "https://api.propublica.org/congress/v1/115/both/bills/enacted.json",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    isLoading={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <section
        className="bill-container"
      >
        <div
          className="bill-msg"
        >
          <h3>
            Below are the bills that became law during the 115th Congress.
          </h3>
        </div>
        <div
          className="cardContainer"
        >
          <Loading />
        </div>
      </section>,
      "className": "bills-pg",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="bill-msg"
          >
            <h3>
              Below are the bills that became law during the 115th Congress.
            </h3>
          </div>,
          <div
            className="cardContainer"
          >
            <Loading />
          </div>,
        ],
        "className": "bill-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h3>
              Below are the bills that became law during the 115th Congress.
            </h3>,
            "className": "bill-msg",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Below are the bills that became law during the 115th Congress.",
            },
            "ref": null,
            "rendered": "Below are the bills that became law during the 115th Congress.",
            "type": "h3",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Loading />,
            "className": "cardContainer",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "section",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <section
          className="bill-container"
        >
          <div
            className="bill-msg"
          >
            <h3>
              Below are the bills that became law during the 115th Congress.
            </h3>
          </div>
          <div
            className="cardContainer"
          >
            <Loading />
          </div>
        </section>,
        "className": "bills-pg",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="bill-msg"
            >
              <h3>
                Below are the bills that became law during the 115th Congress.
              </h3>
            </div>,
            <div
              className="cardContainer"
            >
              <Loading />
            </div>,
          ],
          "className": "bill-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h3>
                Below are the bills that became law during the 115th Congress.
              </h3>,
              "className": "bill-msg",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Below are the bills that became law during the 115th Congress.",
              },
              "ref": null,
              "rendered": "Below are the bills that became law during the 115th Congress.",
              "type": "h3",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Loading />,
              "className": "cardContainer",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "section",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
